misc: build with macOS

roarkanize opened the pull request:
this pull request makes the changes needed to allow bakelite to build in macOS successfully, namely:
- introducing a new file `endian.h` which does the proper macros to supply the expected functions in macOS, or simply includes `<endian.h>` in other platforms
- including the `<sys/random.h>` header on macOS for all the files using `getentropy(2)`
- defining `pipe2` as a function on macOS for `store.c`, since it's not supplied by the standard library
- defining the expected fields to `struct stat` as macros, since they have different names on macOS

changes have been tested in both Linux (Fedora 35) and macOS (macOS 12.3.1) and it builds, backups and restores successfully in both platforms

(I am aware that defining `pipe2` like this can result in race conditions, but it's the only solution, sadly)

Commit:
misc: build with macOS

richfelker commented on backup.c:
```diff
@@ -18,6 +18,12 @@
+#if defined(__APPLE__) || defined(__NetBSD__) 
+#define st_atim st_atimespec
+#define st_ctim st_ctimespec
+#define st_mtim st_mtimespec
```
This is just Apple being gratuitously intentionally wrong. The standard names are `st_[acm]tim`.

richfelker commented on bloom.c:
```diff
@@ -1,11 +1,15 @@
 #include "bloom.h"
 
+#ifdef __APPLE__
+#include <sys/random.h>
```
`getentropy` is approved for future editions of the standard to appear in `unistd.h` which was also the original BSD location. It's odd that Apple is doing this gratuitously wrong.

richfelker commented on endian.h:
```diff
@@ -0,0 +1,20 @@
+#define htobe64(x) OSSwapHostToBigInt64(x)
+#define htole64(x) OSSwapHostToLittleInt64(x)
+#define be64toh(x) OSSwapBigToHostInt64(x)
+#define le64toh(x) OSSwapLittleToHostInt64(x)
```
`endian.h` is also a future standard and a near-universal existing practice. If we're going to have a fallback for systems that lack it, it should be done by detection and with portable versions of the macros, not `#if defined(__APPLE__)` and definitions using the nonstandard Apple functions.

richfelker commented on store.c:
```diff
@@ -13,6 +13,16 @@
+	return 0;
+}
+#endif
+
```
`pipe2` is also future standard, but the above should be conditioned on `!defined(HAVE_PIPE2)` (with detection for that) and in its own file. It should also be using the flags rather than unconditionally setting close-on-exec, and using the right flag bit `FD_CLOEXEC` not `O_CLOEXEC`. But as long as we're not multithreaded we could just always use a "fake" version of `pipe2` like the above without conditioning it on detection that `pipe2` is missing.

richfelker reviewed (commented):
I'd really like to not go down the 1980s-style rabbithole of quirky-OS-specific hard-coded `#ifdef`s. A few of these things are hard bugs that really need to be fixed on their side, or by interposing a set of compatibility headers to fix their stuff. Some of them are extension (and mostly future-standard) functionality that we should perhaps be providing fallbacks for. See comments on particular changes.

roarkanize commented:
Maybe a sort of `configure` script looking for these specific features would be a better solution instead? I'm currently installing a couple virtual machines to test on other platforms as well, so maybe I could expand the scope of this to encompass more platforms than just macOS

Commit:
Revert "misc: build with macOS"

This reverts commit a86831aaae981c486defb4403efa6286ceee7d22.

Commit:
build: add `configure` script

tested with Linux, macOS, FreeBSD and OpenBSD

Commit:
build: add support for NetBSD

roarkanize commented:
A portable implementation for the `endian.h` macros and for `getentropy(2)` are still pending, but I was wondering, is it really worth it, considering (nearly) all UNIX-like OSes have them?

richfelker commented:
For `getentropy`, there is no portable implementation or safe fallback that widely works. What we can easily do though is just declare it ourselves (`int getentropy(void*, size_t);`) so as not to care if `unistd.h` didn't declare it. Then, if the system doesn't have it, you need to provide a library giving a safe version via `LDFLAGS` (but AIUI OSX does have it so that's a non-issue).

Commit:
build: add portable `endian.h` macros implementation

add portable `<endian.h>` macros implementation and
endianness detection at configure phase

add `config.h` to .gitignore

add ´getentropy´ definition to `<unistd.h>` as per comments in #5

Commit:
Merge branch 'richfelker:main' into main
